File: UCF HSPT Project/src/y2013/craft.java

// 27th Annual UCF High School Programming Tournament
//
// Problem Name: Crafty Trolling (craft)
//
// Author:        Skyler Goodell
// Judge Data:    Nick Buelich
// C Solution:    Antony Stabile
// Java Solution: Andrew Harn
// Verifier:      Aaron Teetor

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayDeque;
import java.util.Scanner;


public class craft {

   //Delta array to direct where the lava goes
   int[][] delta = new int[][] {
         {0, 0, 1},
         {0, 1, 0},
         {0, 0 , -1},
         {0, -1, 0}
   };

   //Constructor - Holds the solution
   public craft() {

      //Open up the file
      Scanner sc = null;
      try {
         sc = new Scanner(new File("craft.in"));
      } catch (FileNotFoundException e) {
      }

      //Get the number of maps
      int numberOfMaps = sc.nextInt();

      //Process each map
      for (int mapNumber = 1; mapNumber <= numberOfMaps; mapNumber++) {

         //Get the dimensions of the map
         int l = sc.nextInt(),
             w = sc.nextInt(),
             h = sc.nextInt();

         //This stores our answer
         int lava = 0;

         //Queue of lava block coordinates
         ArrayDeque<int[]> q = new ArrayDeque<int[]>();

         //Create the map and visited array
         char [][][] map = new char[h][l][];
         boolean [][][] visited = new boolean[h][l][w];

         //Read in the map
         for (int i = 0; i < h; i++) {

            for (int j = 0; j < l; j++) {
               map[i][j] = sc.next().toCharArray();

               //Find the lava block and add it to our queue of lava
               for (int k = 0; k < w; k++) {

                  if (map[i][j][k] == '*') {

                     //And add them to the queue
                     lava++;
                     visited[i][j][k] = true;
                     q.add(new int[]{i, j, k});
                  }   
               }
            }
         }

         //Run a Breadth First Search (BFS) to see where the lava will spread
         while (!q.isEmpty()) {

            //Get the next block and its coordinates
            int[] nextBlock = q.poll();

            int x = nextBlock[0],
                y = nextBlock[1],
                z = nextBlock[2];

            //See if we can move down
            if (x < h - 1 && map[x + 1][y][z] == '.') {

               //And we haven't visited that space before
               if (!visited[x + 1][y][z]) {
                  //If so, move down and add the new block to the queue
                  lava++;
                  visited[x + 1][y][z] = true;
                  q.add(new int[]{x + 1, y, z});
               }
            }
            //Otherwise, spread horizontally
            else {
               for (int i = 0; i < 4; i++) {

                  int newX = x + delta[i][0],
                      newY = y + delta[i][1],
                      newZ = z + delta[i][2];

                  //Make sure it's in the board
                  if (newY >= 0 && newY < l && newZ >= 0 && newZ < w) {
                     //Then see if it's empty
                     if (map[newX][newY][newZ] == '.' && 
                         !visited[newX][newY][newZ] ) {
                        //If so, move down and add the new block to the queue
                        lava++;
                        visited[newX][newY][newZ] = true;
                        q.add(new int[]{newX, newY, newZ});
                     }
                  }
               }
            }
         }

         //Lastly, print out the number of lava blocks
         System.out.println("Map #" + mapNumber + ": " + lava);
      }

      // Close file
      sc.close();
   }

   //Main method
   public static void main(String args[]) {
      //Solution written in the constructor
      new craft();
   }
}

File: UCF HSPT Documents/2013/Solutions/craft.c

// 27th Annual UCF High School Programming Tournament
//
// Problem Name: Crafty Trolling (craft)
//
// Author:        Skyler Goodell
// Judge Data:    Nick Buelich
// C Solution:    Antony Stabile
// Java Solution: Andrew Harn
// Verifier:      Aaron Teetor


/*
Strategy:

For each test case, do the following
1) Read in the grid, and locate the lava source
2) Perform a flood fill using a Depth First Search (DFS)
3) Count the number of squares with lava
*/


#include <stdio.h>
#include <stdlib.h>

#define MAXGRID 60

#define EMPTY '.'
#define LAVA '*'
#define WALL '#'

int isOk(int x, int y, int z);
void fill(int x, int y, int z);

// the map and dimensions
char board[MAXGRID][MAXGRID][MAXGRID];
int h,l,w;

int main() {
    FILE * infile;
    int numCases, curCase, i, j, k;
    int sx, sy, sz;

    // open input file
    infile = fopen("craft.in", "r");

    // read in the number of cases
    fscanf(infile, "%d ", &numCases);

    // loop through each case
    for(curCase = 1; curCase <= numCases; curCase++) {
        // read in the dimensions
        fscanf(infile, "%d %d %d ", &l, &w, &h);

        // read in h slices of l lines
        for(i=0; i<h; i++)
            for(j=0; j<l; j++)
                fscanf(infile, "%s ", board[i][j]);

        // locate the source and fill from there
        for(i=0; i<h; i++)
            for(j=0; j<l; j++)
                for(k=0; k<w; k++)
                    if(board[i][j][k] == LAVA) {
                        sx = i;
                        sy = j;
                        sz = k;
                    }

        // fill the board
        board[sx][sy][sz] = EMPTY;
        fill(sx,sy,sz);

        // count the number of lava squares
        int count = 0;
        for(i=0; i<h; i++)
            for(j=0; j<l; j++)
                for(k=0; k<w; k++)
                    if(board[i][j][k] == LAVA)
                        count++;

        // print the answer
        printf("Map #%d: %d\n", curCase, count);
    }

    // exit
    fclose(infile);
    return 0;
}

// performs a recursive dfs from (x,y,z)
// if possible, lava flows down
// otherwise, it flows in the other 4 directions
void fill(int x, int y, int z) {
    // check if we exited the board dimensions
    if(!isOk(x,y,z)) return;

    // if we already filled this square or hit a wall, return
    if(board[x][y][z] != EMPTY) return;

    // fill this square with lava
    board[x][y][z] = LAVA;

    if(!isOk(x+1,y,z) || board[x+1][y][z] == WALL) {
        // we can't go down so spread in the other four directions
        fill(x,y-1,z);
        fill(x,y+1,z);
        fill(x,y,z-1);
        fill(x,y,z+1);
    }
    else {
        // it seems okay to flow down, so do it
        fill(x+1,y,z);
    }
}


// checks if (x,y,z) lies within the board
int isOk(int x, int y, int z) {
    return x>=0 && y>=0 && z>=0 && x<h && y<l && z<w;
}

